runtime/lua/vim/diagnostic.lua

```diff
@@ -890,22 +872,34 @@ M.handlers.signs = {
     end
 
     local ns = M.get_namespace(namespace)
-    if not ns.user_data.sign_group then
-      ns.user_data.sign_group = string.format('vim.diagnostic.%s', ns.name)
+    if not ns.user_data.sign_ns then
+      ns.user_data.sign_ns = api.nvim_create_namespace('')
     end
 
-    local sign_group = ns.user_data.sign_group
+    opts.signs.text = vim.tbl_extend(
+      'force',
+      vim
+        .iter({ unpack(M.severity, 1, 4) })
+        :map(function(v)
+          return v:sub(1, 1):upper()
+        end)
+        :totable(),
+      opts.signs.text or {}
+    )
```

Comment:
This modifies the `opts` table, which we do not want. Better to assign to a new `local` table.

Also, maybe it's just me but I find this `iter` block to be pretty opaque. A simple for loop (as I suggested originally) might be a win here for readability.

```lua
local text = {}
for k in pairs(vim.diagnostic.severity) do
  if opts.sign.text[k] then
    text[k] = opts.sign.text[k]
  elseif type(k) == 'string' and not text[k] then
    text[k] = string.sub(k, 1, 1):upper()
  end
end
```

Reply:
Please correct me if I'm wrong. I haven't looked at the assembly code of lua for. Doesn't this add all M.severity to the for loop? We only need 4.

>  might be a win here for readability.

haha I most like fp style recently

Reply:
```lua
 elseif type(k) == 'string' and not text[k] then
    text[k] = string.sub(k, 1, 1):upper()
  end
```
this is redundant . we just use number type index for match `text[diagnostic.severity] or 'U'`

Reply:
In any case, `{ unpack(M.severity, 1, 4) }` is too clever, it's not at all obvious what this does or why `1` and `4` are used, and makes an assumption on what the values of each of the severities are. Writing out

```lua
{
  [vim.diagnostic.severity.ERROR] = 'E',
  [vim.diagnostic.severity.WARN] = 'W',
  [vim.diagnostic.severity.INFO] = 'I',
  [vim.diagnostic.severity.HINT] = 'H',
}
```

is more verbose, but much more clear. There's no reason to be terse here if it sacrifices clarity.

Reply:
Is `vim.diagnostic.severity` not redundant here? The config interface itself is already `vim.diagnostic.signs`. I would just make it a severity key -> text mapping like I [suggested](https://github.com/neovim/neovim/pull/26193#discussion_r1404873099) earlier.

Reply:
>Is vim.diagnostic.severity not redundant here?

Yes, we can use `M.severity` instead.